README.MD
#  概述
- 用airtest做移动端自动化
- 支持安卓，ios,web
- 基于yml编写用例


## 开始运行
- 编写配置文件
```buildoutcfg
[default]
# 包名
pkg = com.jianshu.haruki
# 手机名字
phone=雷电模拟器
# 设备连接名字
dev_connect=Android://127.0.0.1:5037/
# 设备id,adb devices 获取，如果启动多设备，这里无效
dev=emulator-5554
# 用例目录
root_path=E:\proj\airtest_auto\air_case\android
 # test_plan=1 表示调试用例需要配合test_module使用，0表示全部用例;如果启动多设备,test_plan默认为0,设置1也无效
test_plan=1
test_module=["他的"]
# false|true 表示是否运行用例之前，删除log文件夹，删除后会影响查看历史报告
remove_log=true
# false|true 是否开启录屏,若传true就会自动录屏，但是在模拟器上录屏失败，只使用于真机
recording=false
# android|ios|web，当填如web时，需要填写驱动文件位置
platform =android
# driver_path=E:\proj\airtest_auto\exe\chromedriver.exe
driver_path=
# 服务器信的信息，比如本地的ip，用来展示报告
report_host=172.31.105.196
# 本地服务器路径，我把源代码中的report也放在里面了
local_host_path=E:\proj\aritest
# 本地服务器端口
local_host_port=8000
# 启动什么模式，单机和多机
# 如果指定multi就是多机并行，读取multi的配置，否则就为单机模式
boot=multi
[multi]
# 给不同的设备分配测试模块,dev中的设备id一定要填对
test_case = [{"dev":"ZL9LC685V86DNNMN","test_module": ["我的"],"phone":"真机1"}, {"dev":"emulator-5554","test_module":["他的"],"phone":"雷电"}]
#test_case = [{"dev":"emulator-5554","test_module":["他的"],"phone":"雷电"}]

```

## 测试报告
![summary.png](img/summary.png "report2.png")
![detail.png](img/detail.png "report2.png")
 
   
## 其他
[查看我的changelog](changelog.MD)


changelog.MD
# 2022-1-24
- 使用自带的http.server造成阻塞，修改为Twisted
- 修改多机并行，安卓真机和雷电模拟器测试通过
- 适配单机和多设备web测试，本次测试为谷歌浏览器

# 2022-1-20
- 新增测试报告查看采用本地服务器查看，本次代码在runner5.py
- 测试报告和项目剥离
- 删除其他不用的启动器代码
- 由于修改了report.py源代码，也已经拷贝到根目录可供参考


# 2021-5-15
- 修复录屏不展示问题
- 新增删除历史html报告


# 2021-4-22
- 本次使用雷电模拟器
- 使用夜神模拟器无法截图，连接airtestide出现黑屏
- 打开夜神模拟器的开发者模式-usb调试，adb deivces就能获取到设备列表，若无法获取请重启模拟器
- 使用config/setting.ini进行配置,理论上支持配置ios
- 本次使用代码入口是runner0.py

    


# 2021-3-6 
- runner1.py新增发送压缩测试报告到指定邮箱
- 新增是否删除log文件字段，若删除会影响到历史报告查看
# 2020-9-26 
- runner1.py中新增summary.html给CI使用
- 优化手机录屏失败问题
- 剥离测试报告，摆脱本地依赖
- windows的CI简单运用
# 2020-8-29
- 优化日志记录方式
- 引入selenium，在runner4.py

# 2020-7-12
- 新增多设备并行，代码在runner3.py中
- config/case_data.json 记录测试设备和测试手机
- result_data 记录所有测试用例运行情况
- 修复某些手机测试报告无录像回放

# 2020-7-5
## 调整逻辑
- 重构所有代码
- 读取用例模式改变，修改为testcae/module/test.air的方式
- ```runner1.py```
```buildoutcfg
  # test_plan=1 表示调试用例，运行某几个模块用例；0表示全部用例，运行全部模块用例
    data = {"root_path": root_path, "test_plan": 0, "test_module": [], "dev": "TPG5T18130013404", "phone": "Nova2s"}
    # data = {"root_path": root_path, "test_plan": 1, "test_module": ["他的"], "dev": "TPG5T18130013404", "phone": "Nova2s"}
    run_case(data)
```
- 新增调试单个模块下用例
- ```runner2.py```
```buildoutcfg
    root_path = PATH("air_case/我的")
    test_plan = 1  # 0表示运行模块下所有用例,1表示运行模块下的test_data里面的用例
    test_data = ["打开我的文章.air"]
    run_case({"root_path": root_path, "test_plan": test_plan, "test_data": test_data,  "dev": "TPG5T18130013404",
              "phone": "Nova2s"})
```
## 注意事项
- **windows中用pycharm若出现adb命令异常**
    - 如无法获取到设备尽量需要和airtestIDE的adb保持一致，详细参考[官方文档](https://airtest.netease.com/docs/cn/2_device_connection/2_android_faq.html)
- **测试详细报告中，如果没有视频回放**
    - 手动打开手机端的yosemite，点击开始录屏，给予权限，最好是和PocoService一样，用手动管理，所有启动全部打开
- **想同时允许多个设备**
    - 只要在runner1/runner2中，修改不同的参数，然后在pycharm中多次运行，就能启动多设备
    - 后续新增多进程多机并发
# 2020-6-12
- 代码在runner2.py
- 支持用例采用yml编写方式
    - 自定义，如：不用再到每个用例步骤填写等待时间
    - 灵活
    - 容错
- 支持用例文件夹下读取模块下用例，如testcase/me/case001_me_打开个人中心.air
- 测试报告向新增时间
### 待完成
- 测试报告详情新增视频回放

# 2020-4
## 1.0 特性
- runner.py
    - 支持安卓批量执行用例
    - 执行失败重测
    - 支持指定部分用例进行调试
    - 支持只运行失败用例
- runner1.py
    - 支持多机并行

runner3.py
#!/usr/bin/env python
# -*- coding=utf-8 -*-
import datetime

__CreateAt__ = '2020/4/19-17:34'

import threading

from multiprocessing import Pool

from airtest.cli.runner import AirtestCase, run_script
from argparse import *
import airtest.report.report as report
import jinja2
import shutil
import os
import io

from http_server import HttpServer
from read_ini import ReadIni
from util.android_util import attached_devices
from util.common import get_test_case_runner2, get_case_total_time, get_test_modules, Runner3Common
from datetime import datetime
import time

PATH = lambda p: os.path.abspath(
    os.path.join(os.path.dirname(__file__), p)
)


def run_case(data):
    dev_connect = None
    # 当平台为安卓时，检查是否连接成功
    if data.get("platform", "1") == "android":
        print("启动安卓启动器")
        devices = attached_devices()
        if not devices:
            print("无可用设备")
            return
        dev_connect = data["dev_connect"]
    elif data.get("platform", "1") == "ios":
        pass
    elif data.get("platform", "1") == "web":
        print("启动web启动器")
    test = CustomAirtestCase(data["root_path"], dev_connect)
    test.run_air(data)


def run(root_dir, test_case, device, dev_connect, local_host_path, log_date, recording, report_host, phone):
    if device == "web":
        air_device = None
    else:
        air_device = [dev_connect + device]  # 取设备
    script = os.path.join(root_dir, test_case["module"], test_case["case"])
    log_host_path = os.path.join(test_case["module"], test_case["case"].replace('.air', ''))
    log = os.path.join(local_host_path, log_host_path)
    os.makedirs(log)
    print(str(log) + ' 创建日志文件成功')
    output_file = os.path.join(log, 'log.html')
    args = Namespace(device=air_device, log=log, compress=None, recording=recording, script=script, no_image=None)
    # 用例开始执行日期
    st_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    # 用例开始执行时间
    s_time = datetime.now().strftime("%H:%M:%S")
    try:
        run_script(args, AirtestCase)
        is_success = True
    except:
        is_success = False
    end_date = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    # 用例结束执行时间
    e_time = datetime.now().strftime("%H:%M:%S")
    # 用例耗时时间
    sum_time = get_case_total_time(s_time, e_time)
    # 生成测试用例的详情报告
    rpt = report.LogToHtml(script, log, report_host + "/report",
                           log_host=report_host + "/log/" + log_date + "/" + log_host_path)

    rpt.report("log_template.html", output_file=output_file)
    # 记录测试结果
    s_name = test_case["case"].replace(".air", "")
    s_log = os.path.join(report_host, "log", log_date, test_case["module"], s_name)
    result = {"result": is_success, "start_date": st_date, "end_date": end_date, "sum_time": sum_time,
              "log": s_log, "name": s_name, "phone": phone, "dev": device}

    Runner3Common.set_result_json(result)

    if not is_success:
        print("存在失败用例")
        print(test_case)


class CustomAirtestCase(AirtestCase):
    # @classmethod
    # def setUpClass(cls):
    #     super(CustomAirtestCase,cls).setUpClass()

    def __init__(self, root_path, dev_connect):
        self.fail_data = []
        self.root_path = root_path  # 用例目录
        self.dev_connect = dev_connect  # 设备连接 如：Android://127.0.0.1:5037/
        self.log_date = ""  # 报告的跟目录 /log/201201201201
        self.local_host_path = ""  # 服务器报告目录
        self.recording = ""  # 是否录屏
        self.report_host = ""  # 服务器远程地址，如http://172.31.105.196:8000
        self.dev = "设备id号" # adb device 获取
        super().__init__()

    def setUp(self):
        print("custom setup")
        super(CustomAirtestCase, self).setUp()

    def tearDown(self):
        print("custom tearDown")
        # exec teardown script
        # self.exec_other_script("teardown.owl")
        super(CustomAirtestCase, self).setUp()

    def run_case1(self, data_item):
        self.dev = data_item["dev"]
        # 记录测试设备
        Runner3Common.set_case_module_dev({"test_dev": self.dev})
        # 记录测试模块
        for i in data_item["test_module"]:
            Runner3Common.set_case_module_dev({"test_modules": i})

        get_case_data = Runner3Common.get_cases(data_item, self.dev, self.root_path)
        # 循环执行用例
        for i in get_case_data:
            run(root_dir=self.root_path, test_case=i, device=self.dev, dev_connect=self.dev_connect,
                local_host_path=self.local_host_path,log_date=self.log_date,
                recording=self.recording, report_host=self.report_host, phone=data_item["phone"])

    def run_air(self, data):
        self.recording = data["recording"]
        self.report_host = "http://" + data["report_host"] + ":" + data["local_host_port"]
        # 开启本地http服务器
        thread_http = threading.Thread(target=HttpServer.start, args=(),
                         kwargs={"local_host_path": data["local_host_path"], "port": data["local_host_port"]})
        thread_http.start()
        time.sleep(2)
        # 日志按照日期格式生成
        self.log_date = datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
        # 生成本地的测试目录
        log_path = os.path.join(data["local_host_path"], "log")
        self.local_host_path = os.path.join(log_path, self.log_date)
        # remove_log表示如果传值就会删除整个log文件夹，删除后无法查看历史报告
        if data["remove_log"] and os.path.isdir(log_path):
            shutil.rmtree(log_path)

        os.makedirs(self.local_host_path)
        print(str(self.local_host_path) + ' 日志根目录创建')

        # 预留位置1：此处可以放卸载安装应用
        # 预留位置2：获取手机验证码，登录；可以采用打开手机短信应用获取后，输入到验证码，最好是用一个万能验证码

        # 整个用例开始执行时间
        start_time = datetime.now().strftime("%H:%M:%S")
        pool = Pool()
        pool.map(self.run_case1, data["test_case"])
        # 整个用例结束执行时间
        end_time = datetime.now().strftime("%H:%M:%S")
        # 以小时，分钟，秒钟的方式记录所有用例耗时时间
        total_time = get_case_total_time(start_time, end_time)
        test_modules_dev = Runner3Common.get_case_module_dev(PATH("config/case_data.json"))
        print("==设备=%s" % test_modules_dev)

        # 计算用例成功，失败,总数
        get_case_data = Runner3Common.get_result_json()
        success = 0
        s_count = len(get_case_data["data"])
        for j in get_case_data["data"]:
            if j.get("result"):
                success += 1

        # 记录用例结果
        Runner3Common.set_result_summary_json({"start_time": self.log_date, "success": success, "count": s_count,
                                               "total_time": total_time, "phone": test_modules_dev["test_dev"],
                                               "modules": test_modules_dev["test_modules"], "dev": self.dev})

        # 生成测试报告
        env = jinja2.Environment(
            loader=jinja2.FileSystemLoader(self.local_host_path),
            extensions=(),
            autoescape=True
        )
        print("测试结果:%s" % Runner3Common.get_result_json())
        # 先删除模板文件，然后拷贝进去，防止模板文件更新不及时
        t = os.path.join(self.local_host_path, "multi_summary_template.html")
        if os.path.exists(t):
            shutil.rmtree(t)
        shutil.copy(PATH("util/multi_summary_template.html"), self.local_host_path)

        template = env.get_template("multi_summary_template.html", self.local_host_path)
        html = template.render({"results": Runner3Common.get_result_json()})
        put_html = os.path.join("summary_%s.html" % datetime.now().strftime("%Y-%m-%d-%H-%M-%S"))
        output_file = os.path.join(self.local_host_path, put_html)
        with io.open(output_file, 'w', encoding="utf-8") as f:
            f.write(html)
        # 按日期生成测试报告,方便对比历史报告，但是程序入口字段需要设置为"remove_log": False
        print("测试报告本地路径为：%s" % output_file)
        print("测试报告访问服务器为：%s/%s/%s/%s" % (self.report_host, "log", self.log_date, put_html))
        thread_http.join()

def multi_runner(data1):
    fun_path = lambda p: os.path.abspath(
        os.path.join(os.path.dirname(__file__), p)
    )
    # 初始化记录数据的json
    Runner3Common.init_result(fun_path("config/result_data.json"), fun_path("config/case_data.json"))

    run_case(data1)


if __name__ == '__main__':
    pass
    # PATH = lambda p: os.path.abspath(
    #     os.path.join(os.path.dirname(__file__), p)
    # )
    # path = PATH("config/setting1.ini")
    # data1 = ReadIni(path).get_ini_list()
    # if data1["boot"] == "multi":
    #     # 初始化记录数据的json
    #     Runner3Common.init_result(PATH("config/result_data.json"), PATH("config/case_data.json"))
    #     run_case(data1)
    # else:
    #     pass
